client/matching/isolation_loadbalancer.go

```diff
@@ -72,71 +76,50 @@ func (i *isolationLoadBalancer) PickWritePartition(taskListType int, req WriteRe
 
 func (i *isolationLoadBalancer) PickReadPartition(taskListType int, req ReadRequest, isolationGroup string) string {
 	taskList := *req.GetTaskList()
-	nRead := i.provider.GetNumberOfReadPartitions(req.GetDomainUUID(), taskList, taskListType)
-	taskListName := taskList.Name
 
-	if nRead <= 1 {
-		return taskListName
+	domainName, err := i.domainIDToName(req.GetDomainUUID())
+	if err != nil || !i.isolationEnabled(domainName) {
+		return i.fallback.PickReadPartition(taskListType, req, isolationGroup)
 	}
 
-	partitions, ok := i.getPartitionsForGroup(isolationGroup, nRead)
+	config := i.provider.GetPartitionConfig(req.GetDomainUUID(), taskList, taskListType)
+
+	partitions, ok := i.getPartitionsForGroup(isolationGroup, config.ReadPartitions)
 	if !ok {
 		return i.fallback.PickReadPartition(taskListType, req, isolationGroup)
 	}
 
-	// Scaling down, we need to consider both sets of partitions
-	if numWrite := i.provider.GetNumberOfWritePartitions(req.GetDomainUUID(), taskList, taskListType); numWrite != nRead {
-		writePartitions, ok := i.getPartitionsForGroup(isolationGroup, numWrite)
-		if ok {
-			for p := range writePartitions {
-				partitions[p] = struct{}{}
-			}
-		}
-	}
-
 	p := i.pickBetween(partitions)
 
 	return getPartitionTaskListName(taskList.GetName(), p)
 }
 
 func (i *isolationLoadBalancer) UpdateWeight(taskListType int, req ReadRequest, partition string, info *types.LoadBalancerHints) {
+	i.fallback.UpdateWeight(taskListType, req, partition, info)
 }
 
-func (i *isolationLoadBalancer) getPartitionsForGroup(taskGroup string, partitionCount int) (map[int]any, bool) {
+func (i *isolationLoadBalancer) getPartitionsForGroup(taskGroup string, partitions map[int]*types.TaskListPartition) (map[int]any, bool) {
 	if taskGroup == "" {
 		return nil, false
 	}
-	isolationGroups := slices.Clone(i.allIsolationGroups())
-	slices.Sort(isolationGroups)
-	index := slices.Index(isolationGroups, taskGroup)
-	if index == -1 {
-		return nil, false
-	}
-	partitions := make(map[int]any, 1)
-	// 3 groups [a, b, c] and 4 partitions gives us a mapping like this:
-	// 0, 3: a
-	// 1: b
-	// 2: c
-	// 4 groups [a, b, c, d] and 10 partitions gives us a mapping like this:
-	// 0, 4, 8: a
-	// 1, 5, 9: b
-	// 2, 6: c
-	// 3, 7: d
-	if len(isolationGroups) <= partitionCount {
-		for j := index; j < partitionCount; j += len(isolationGroups) {
-			partitions[j] = struct{}{}
+
+	res := make(map[int]any)
```

Comment:
nit: maybe add some comments. It took a lot of time to figure out what this is doing.

Reply:
I rewrote this to be more readable, good callout.